File: typescript/core/src/promptCache.ts
/**
 * Prompt caching — the static/dynamic system-prompt split. The TypeScript port
 * of the Rust reference `smooth-operator-core::conversation::PromptCache`.
 *
 * A system prompt has two halves with very different churn rates: role
 * instructions and tool schemas barely change, while project context
 * (AGENTS.md / CLAUDE.md, the working set) changes every turn. Anthropic's
 * prompt cache keys on a PREFIX, so putting the volatile half first invalidates
 * the whole thing. {@link PROMPT_CACHE_BOUNDARY} splits them: everything above
 * the marker is static and hashed once for cache-key dedup, everything below is
 * dynamic and can be swapped without busting the static prefix.
 *
 * Feed the result to the agent as its instructions:
 *
 * ```ts
 * const cache = new PromptCache(`${rules}${PROMPT_CACHE_BOUNDARY}${projectContext}`);
 * new SmoothAgent(provider, { instructions: cache.fullPrompt() });
 * ```
 */

/**
 * Marker that splits a system prompt into a cacheable static portion and a
 * frequently-changing dynamic portion.
 */
export const PROMPT_CACHE_BOUNDARY = '__PROMPT_CACHE_BOUNDARY__';

/** A system prompt split at {@link PROMPT_CACHE_BOUNDARY}. */
export class PromptCache {
    /** The cacheable half (above the marker). */
    readonly staticPortion: string;
    /** The frequently-changing half (below the marker). */
    dynamicPortion: string;
    private readonly _staticHash: string;
    private readonly _staticTokens: number;

    /**
     * Split a system prompt at the boundary marker. With no marker the entire
     * prompt is treated as dynamic — nothing is claimed cacheable that the
     * caller didn't mark.
     */
    constructor(prompt: string) {
        const idx = prompt.indexOf(PROMPT_CACHE_BOUNDARY);
        if (idx < 0) {
            this.staticPortion = '';
            this.dynamicPortion = prompt;
        } else {
            this.staticPortion = prompt.slice(0, idx);
            this.dynamicPortion = prompt.slice(idx + PROMPT_CACHE_BOUNDARY.length);
        }
        this._staticHash = hashPromptPortion(this.staticPortion);
        this._staticTokens = this.staticPortion === '' ? 0 : Math.floor(this.staticPortion.length / 4) + 1;
    }

    /**
     * Reassemble static + boundary + dynamic. With no static portion the dynamic
     * half is returned alone, so a prompt that was never split round-trips
     * unchanged rather than gaining a stray marker.
     */
    fullPrompt(): string {
        if (this.staticPortion === '') return this.dynamicPortion;
        return `${this.staticPortion}${PROMPT_CACHE_BOUNDARY}${this.dynamicPortion}`;
    }

    /**
     * Swap the dynamic half, leaving the static half and its hash untouched —
     * the whole point of the split.
     */
    updateDynamic(dynamic: string): void {
        this.dynamicPortion = dynamic;
    }

    /**
     * Identifies the static portion for cache-key deduplication.
     *
     * Process-local only: it is compared against other hashes from THIS engine,
     * never sent on the wire, so it deliberately does not match the Rust
     * reference's value (Rust uses `DefaultHasher`, which is not reproducible
     * across languages — or even across Rust releases). The ported contract is
     * the behavior: same static text hashes the same, different static text
     * hashes differently, and `updateDynamic` never changes it.
     */
    staticHash(): string {
        return this._staticHash;
    }

    /** Estimated tokens the static portion saves on a cache hit. */
    cachedTokens(): number {
        return this._staticTokens;
    }
}

/**
 * FNV-1a (64-bit), the same non-cryptographic hash family the vector embedder
 * uses, rendered as 16 hex chars like the Rust reference. BigInt because a
 * 64-bit multiply overflows a JS number.
 */
function hashPromptPortion(s: string): string {
    const MASK = (1n << 64n) - 1n;
    let h = 14695981039346656037n;
    const bytes = new TextEncoder().encode(s);
    for (const b of bytes) {
        h = (h ^ BigInt(b)) & MASK;
        h = (h * 1099511628211n) & MASK;
    }
    return h.toString(16).padStart(16, '0');
}

File: go/core/prompt_cache.go
package core

// Prompt caching — the static/dynamic system-prompt split. The Go port of the
// Rust reference `smooth-operator-core::conversation::PromptCache`.
//
// A system prompt has two halves with very different churn rates: role
// instructions and tool schemas barely change, while project context
// (AGENTS.md / CLAUDE.md, the working set) changes every turn. Anthropic's
// prompt cache keys on a PREFIX, so putting the volatile half first invalidates
// the whole thing. PromptCacheBoundary splits them: everything above the marker
// is static and hashed once for cache-key dedup, everything below is dynamic and
// can be swapped without busting the static prefix.
//
// This is the caller-facing half of prompt caching. The wire half — attaching
// `cache_control: {"type":"ephemeral"}` to the strategic prefix boundaries — is
// in openai.go.

import (
	"fmt"
	"strings"
)

// PromptCacheBoundary splits a system prompt into a cacheable static portion
// and a frequently-changing dynamic portion.
const PromptCacheBoundary = "__PROMPT_CACHE_BOUNDARY__"

// PromptCache is a system prompt split at [PromptCacheBoundary].
type PromptCache struct {
	staticPortion  string
	dynamicPortion string
	staticHash     string
	staticTokens   int
}

// NewPromptCache splits a system prompt at the boundary marker. With no marker
// the entire prompt is treated as dynamic — nothing is claimed to be cacheable
// that the caller didn't mark.
func NewPromptCache(prompt string) *PromptCache {
	staticPortion, dynamicPortion := "", prompt
	if idx := strings.Index(prompt, PromptCacheBoundary); idx >= 0 {
		staticPortion = prompt[:idx]
		if after := idx + len(PromptCacheBoundary); after < len(prompt) {
			dynamicPortion = prompt[after:]
		} else {
			dynamicPortion = ""
		}
	}

	tokens := 0
	if staticPortion != "" {
		tokens = len(staticPortion)/4 + 1
	}
	return &PromptCache{
		staticPortion:  staticPortion,
		dynamicPortion: dynamicPortion,
		staticHash:     hashPromptPortion(staticPortion),
		staticTokens:   tokens,
	}
}

// StaticPortion is the cacheable half (above the marker).
func (p *PromptCache) StaticPortion() string { return p.staticPortion }

// DynamicPortion is the frequently-changing half (below the marker).
func (p *PromptCache) DynamicPortion() string { return p.dynamicPortion }

// FullPrompt reassembles static + boundary + dynamic. With no static portion
// the dynamic half is returned alone, so a prompt that was never split round-
// trips unchanged rather than gaining a stray marker.
func (p *PromptCache) FullPrompt() string {
	if p.staticPortion == "" {
		return p.dynamicPortion
	}
	return p.staticPortion + PromptCacheBoundary + p.dynamicPortion
}

// UpdateDynamic swaps the dynamic half, leaving the static half and its hash
// untouched — the whole point of the split.
func (p *PromptCache) UpdateDynamic(dynamic string) { p.dynamicPortion = dynamic }

// StaticHash identifies the static portion for cache-key deduplication.
//
// Process-local only: it is compared against other hashes from THIS engine,
// never sent on the wire, so it deliberately does not match the Rust
// reference's value (Rust uses DefaultHasher, which is not reproducible across
// languages — or even across Rust releases). The contract that is ported is the
// behavior: same static text hashes the same, different static text hashes
// differently, and UpdateDynamic never changes it.
func (p *PromptCache) StaticHash() string { return p.staticHash }

// CachedTokens estimates the tokens the static portion saves on a cache hit.
func (p *PromptCache) CachedTokens() int { return p.staticTokens }

// hashPromptPortion is FNV-1a (64-bit), the same non-cryptographic hash the
// vector embedder uses, rendered as 16 hex chars like the Rust reference.
func hashPromptPortion(s string) string {
	const (
		offset64 uint64 = 14695981039346656037
		prime64  uint64 = 1099511628211
	)
	h := offset64
	for i := 0; i < len(s); i++ {
		h ^= uint64(s[i])
		h *= prime64
	}
	return fmt.Sprintf("%016x", h)
}
